### PR TITLE
feat: add codecv5  and codecv6 for Euclid fork

### PR DESCRIPTION
**Pull request by omerfirmak**
### Purpose or design rationale of this PR

*Describe your change. Make sure to answer these three questions: What does this PR do? Why does it do it? How does it do it?*

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

## Release Notes

- **New Features**
  - Introduced flexible codec versioning mechanism.
  - Added support for new codec versions (V5 and V6).

- **Improvements**
  - Enhanced version handling in batch creation and decoding processes.
  - Updated CRC calculation for chunks with no row consumption.

- **Dependency Updates**
  - Updated Go Ethereum and crypto package versions.

- **Testing**
  - Added test coverage for new codec version scenarios and edge cases in CRC calculations.

### CODE REVIEW

**Review by colinlyguo — APPROVED** [review]
Is the CI fail due to "go mod tidy"?

File: encoding/da.go [hunk 1]
```diff
@@ -649,8 +661,10 @@ func GetCodecVersion(config *params.ChainConfig, blockHeight, blockTimestamp uin
 		return CodecV2
 	} else if !config.IsDarwinV2(blockTimestamp) {
 		return CodecV3
-	} else {
+	} else if !config.IsEuclid(blockTimestamp) {
 		return CodecV4
+	} else {
+		return CodecV6
```
Comment (jonastheis) [inline encoding/da.go:667]: Adding a comment here why `CodecV6` is "skipped" would be great

**Review by jonastheis — DISMISSED** [review]

File: encoding/interfaces.go [hunk 2]
```diff
@@ -91,14 +93,20 @@ func CodecFromVersion(version CodecVersion) (Codec, error) {
...
 	default:
 		return nil, fmt.Errorf("unsupported codec version: %v", version)
 	}
 }
 
 // CodecFromConfig determines and returns the appropriate codec based on chain configuration, block number, and timestamp.
 func CodecFromConfig(chainCfg *params.ChainConfig, startBlockNumber *big.Int, startBlockTimestamp uint64) Codec {
-	if chainCfg.IsDarwinV2(startBlockTimestamp) {
+	if chainCfg.IsEuclid(startBlockTimestamp) {
+		return NewDACodecV6()
```
Comment (colinlyguo) [inline encoding/interfaces.go:108]: same as Jonas's comment.

**Review by colinlyguo — DISMISSED** [review]
Do we want to use a stable version of l2geth in `go.mod`? i.e. Would there be some merge orders?

File: encoding/da.go (hunk 1 above)
Comment (Thegaram) [reply encoding/da.go:667]: Do you mean `CodecV5`?

File: go.mod [hunk 3]
```diff
@@ -4,7 +4,7 @@ go 1.21
 
 require (
 	github.com/agiledragon/gomonkey/v2 v2.12.0
-	github.com/scroll-tech/go-ethereum v1.10.14-0.20241210104312-bdf64cfb39dc
+	github.com/scroll-tech/go-ethereum v1.10.14-0.20250103082839-ea3ec93d8c1e
```
Comment (Thegaram) [inline go.mod:7]: Why bump version, is there any missing feature in the old version?

**Review by Thegaram — DISMISSED** [review]

File: go.mod (hunk 3 above)
Comment (omerfirmak) [reply go.mod:7]: Yeap, IsEuclid was a new addition.

**Review by jonastheis — COMMENTED** [review]
LGTM. Just needs fixing the linter errors

**Review by jonastheis — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]
